examples/baroclinic_instability.jl

```diff
@@ -0,0 +1,145 @@
+    "--dTdz"
+        arg_type=Float64
+        required=true
+        help="Temperature gradient to impose at the bottom [K/m]."
+    "--diffusivity"
+        arg_type=Float64
+        required=true
+        help="Diffusivity κ [m²/s]."
+    "--dt"
+        arg_type=Float64
+        required=true
+        help="Time step in seconds."
+    "--days"
+        arg_type=Float64
+        required=true
+        help="Number of simulation days to run the model."
+    "--output-dir", "-d"
+        arg_type=AbstractString
+        required=true
+        help="Base directory to save output to."
+end
+
+parsed_args = parse_args(s)
+N = parsed_args["resolution"]
+L = parsed_args["length"]
+Q = parsed_args["heat-flux"]
+dTdz = parsed_args["dTdz"]
+κ = parsed_args["diffusivity"]
+
+N = isinteger(N) ? Int(N) : N
+τ = isinteger(L) ? Int(L) : L
+Q = isinteger(Q) ? Int(Q) : Q
+
+dt, days = parsed_args["dt"], parsed_args["days"]
+dt = isinteger(dt) ? Int(dt) : dt
+days = isinteger(days) ? Int(days) : days
+
+base_dir = parsed_args["output-dir"]
+
+filename_prefix = "periodic_baroclinic_N" * string(N) * "_L" * string(L) *
+                  "_Q" * string(Q) * "_dTdz" * string(dTdz) * "_k" * string(κ) *
+                  "_dt" * string(dt) * "_days" * string(days)
+output_dir = joinpath(base_dir, filename_prefix)
+
+if !isdir(output_dir)
+    println("Creating directory: $output_dir")
+    mkpath(output_dir)
+end
+
+# Adding a second worker/proccessor for asynchronous NetCDF output writing.
+using Distributed
+addprocs(1)
+
+# We need to activate the Oceananigans environment on all workers if executing
+# this script from the development repository using "julia --project".
+@everywhere using Pkg
+@everywhere Pkg.activate(".");
+
+@everywhere using Oceananigans
+@everywhere using CuArrays
+@everywhere using Printf
+
+# Physical constants.
+ρ₀ = 1027    # Density of seawater [kg/m³]
+cₚ = 4181.3  # Specific heat capacity of seawater at constant pressure [J/(kg·K)]
+
+# Set more simulation parameters.
+Nx, Ny, Nz = N, N, N
+Lx, Ly, Lz = L, L, 100
+Δt = dt
+Nt = Int(days*60*60*24/dt)
+ν = κ  # This is also implicitly setting the Prandtl number Pr = 1.
+
+# Create the model.
+model = Model(N=(Nx, Ny, Nz), L=(Lx, Ly, Lz), ν=ν, κ=κ, arch=GPU(), float_type=Float64)
+
+# To impose a flux boundary condition, the top flux imposed should be negative
+# for a heating flux and positive for a cooling flux, thus the minus sign.
+Δ = 25
+xC = model.grid.xC
+top_flux = @. (-Q + Δ*cos(2π*xC/Lx)) / (ρ₀ * cₚ)
+top_flux = CuArray(repeat(top_flux, 1, Ny))
+
+# Set boundary conditions.
+model.boundary_conditions.T.z.left = BoundaryCondition(Flux, top_flux)
+model.boundary_conditions.T.z.right = BoundaryCondition(Gradient, dTdz)
+
+# Set initial conditions.
+T_prof = 273.15 .+ 20 .+ dTdz .* model.grid.zC  # Initial temperature profile.
+T_3d = repeat(reshape(T_prof, 1, 1, Nz), Nx, Ny, 1)  # Convert to a 3D array.
```

Comment:
`repeat` is not necessary here. For example:

```julia
julia> a = CuArray(zeros(3, 3, 3))
3×3×3 CuArray{Float64,3}:
[:, :, 1] =
 0.0  0.0  0.0
 0.0  0.0  0.0
 0.0  0.0  0.0

[:, :, 2] =
 0.0  0.0  0.0
 0.0  0.0  0.0
 0.0  0.0  0.0

[:, :, 3] =
 0.0  0.0  0.0
 0.0  0.0  0.0
 0.0  0.0  0.0

julia> b = CuArray(reshape(rand(3), (1, 1, 3)))
1×1×3 CuArray{Float64,3}:
[:, :, 1] =
 0.5867310470464673

[:, :, 2] =
 0.7576453305200208

[:, :, 3] =
 0.5525193734871321

julia> a .= b
3×3×3 CuArray{Float64,3}:
[:, :, 1] =
 0.586731  0.586731  0.586731
 0.586731  0.586731  0.586731
 0.586731  0.586731  0.586731

[:, :, 2] =
 0.757645  0.757645  0.757645
 0.757645  0.757645  0.757645
 0.757645  0.757645  0.757645

[:, :, 3] =
 0.552519  0.552519  0.552519
 0.552519  0.552519  0.552519
 0.552519  0.552519  0.552519
```

Reply:
Ah nice, but don't think it would work here without creating `T_3d` on a separate line. I guess line 107 is also creating `T_3d` while your example requires that `T_3d` already be a `CuArray` of the right size.

Reply:
I see: you want horizontally inhomogeneous perturbations. 

I’ll work on the abstraction that cleans this up soon...

Reply:
Yeah I'm sure it wouldn't be hard to have something like `set_vertical_profile(::Field, ::Vector)` and `add_random_perturbations(...)`, just haven't had the time to prioritize this type of nice user-friendly stuff :(